languages/go/reference/exercise-concepts/parallel-letter-frequency.md

```diff
@@ -0,0 +1,29 @@
+# Concepts of Parallel Letter Frequency
+
+## Concurrency
+
+The goal of this exercise is to learn to write concurrent code to count the letters inside different texts.
+
+## Goroutines
+
+In order to count concurrent the apppearance of different letters in separate texts it is important to be able to create goroutines with the `go` keyword, one for each text where the letters are counted.
+
+## Channel
+
+In this exercise the student can learn to use a channel to make the main goroutine communicate with the goroutines that count letters.
+
+## Functions
+
+The function that has to be called concurrently to count needs to be created in order to get the total count at the end.
+
+### Anonymous functions
+
+However, since it is a simple function it is not necessary to define an external function but it can be coded inside the `go` statement.
+
+## For range
+
+To iterate on the texts and on the frequency map a plain `for` clause is no more needed than a for range statement as there is no specific manipulation to do on the index of the `for` loop.
```

Comment:
I'll sometimes see the index variable (or both variables) ignored with the blank identifier, when this is unnecessary; `for range input` is valid, and elegant.

Reply:
There's a common optimization I see here which is to define the result map as the first receive from the channel, and then loop awkwardly over it with a triple-clause `for` statement:

```go
freq := <-results
for i := 1; i < len(input); i++ {
	for k, v := range <-results {
		...
```

This is ugly and unnecessary, when we can simply write:

```go
var freq = FreqMap{}
for range input {
	for k, v := range <-results {
```

However, students will often fight like cornered wildcats for their favourite performance optimization. If they haven't yet developed the aesthetic sense to see why this isn't a good idea, I'll explain that it makes it harder to understand the code. Worse, the loop statement is dangerously close, but different, to the standard `for i := 0...`, so the reader is almost guaranteed to misunderstand this.

Reply:
The `concepts for v2 exercise`-merge-requests do not target to evaluate bad, good or better patterns. This is not our goal with these documents.

Our goal is to list the concepts needed and commonly used by students to solve an exercise and give a short reason how/why they are used.